std/experimental/color/conv.d

```diff
@@ -0,0 +1,523 @@
+    }
+    if(hex.length == 6)
+    {
+        ubyte r = cast(ubyte)(val(hex[0]) << 4) | val(hex[1]);
+        ubyte g = cast(ubyte)(val(hex[2]) << 4) | val(hex[3]);
+        ubyte b = cast(ubyte)(val(hex[4]) << 4) | val(hex[5]);
+        return cast(Color)RGB8(r, g, b);
+    }
+    if(hex.length == 8)
+    {
+        ubyte a = cast(ubyte)(val(hex[0]) << 4) | val(hex[1]);
+        ubyte r = cast(ubyte)(val(hex[2]) << 4) | val(hex[3]);
+        ubyte g = cast(ubyte)(val(hex[4]) << 4) | val(hex[5]);
+        ubyte b = cast(ubyte)(val(hex[6]) << 4) | val(hex[7]);
+        return cast(Color)RGBA8(r, g, b, a);
+    }
+    else
+    {
+        // TODO: should we look up colors from the W3C color table by name?
+
+        assert(false, "Invalid hex string!");
+    }
+}
+
+unittest
+{
+    // 3 digits
+    static assert(colorFromString("F80") == RGB8(0xFF,0x88, 0x00));
+    static assert(colorFromString("#F80") == RGB8(0xFF,0x88, 0x00));
+    static assert(colorFromString("$F80") == RGB8(0xFF,0x88, 0x00));
+    static assert(colorFromString("0xF80") == RGB8(0xFF,0x88, 0x00));
+
+    // 6 digits
+    static assert(colorFromString("FF8000") == RGB8(0xFF,0x80, 0x00));
+    static assert(colorFromString("#FF8000") == RGB8(0xFF,0x80, 0x00));
+    static assert(colorFromString("$FF8000") == RGB8(0xFF,0x80, 0x00));
+    static assert(colorFromString("0xFF8000") == RGB8(0xFF,0x80, 0x00));
+
+    // 4/8 digita (/w alpha)
+    static assert(colorFromString!RGBA8("#8C41") == RGBA8(0xCC,0x44, 0x11, 0x88));
+    static assert(colorFromString!RGBA8("#80CC4401") == RGBA8(0xCC,0x44, 0x01, 0x80));
+}
+
+
+package:
+
+// convert between pixel data types
+To convertPixelType(To, From)(From v) if(isValidComponentType!From && isValidComponentType!To)
+{
+    import std.algorithm: max;
+
+    static if(isIntegral!From && isIntegral!To)
+    {
+        // extending normalised integer types is not trivial
+        return convertNormInt!To(v);
+    }
+    else static if(isIntegral!From && isFloatingPoint!To)
+    {
+        alias FP = FloatTypeFor!(From, To);
+        static if(isSigned!From) // max(c, -1) is the signed conversion followed by D3D, OpenGL, etc.
+            return To(max(v*FP(1.0/From.max), FP(-1.0)));
+        else
+            return To(v*FP(1.0/From.max));
+    }
+    else static if(isFloatingPoint!From && isIntegral!To)
+    {
+        alias FP = FloatTypeFor!(To, From);
+        // HACK: this is incomplete!
+        //       +0.5 rounding only works for positive numbers
+        //       we also need to clamp (saturate) [To.min, To.max]
+        return cast(To)(v*FP(To.max) + FP(0.5));
+    }
+    else
+        return To(v);
+}
+
+
+// converts directly between fixed-point color types, without doing float conversions
+// ** this should be tested for performance; we can optimise the small->large conversions with table lookups
+To convertNormInt(To, From)(From i) if(isIntegral!To && isIntegral!From)
```

Comment:
IMO this function is also to big and should be split similar to the function above